File: typescript/food-ordering/app/src/restate-app/order_workflow.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

import * as restate from "@restatedev/restate-sdk";
import * as deliveryManager from "./delivery_manager";
import {Order, Status} from "./types/types";
import * as orderstatus from "./order_status_service";
import { completed } from "./utils/utils";
import {getPaymentClient} from "./clients/payment_client";
import {getRestaurantClient} from "./clients/restaurant_client";

/**
 * Order processing workflow Gets called for each Kafka event that is published to the order topic.
 * The event contains the order ID and the raw JSON order. The workflow handles the payment, asks
 * the restaurant to start the preparation, and triggers the delivery.
 */

const restaurant = getRestaurantClient();
const paymentClnt = getPaymentClient();

const eventHandler = async (ctx: restate.RpcContext, event: restate.Event) => {
  const order = JSON.parse(event.json<string>());
  await create(ctx, event.key, order);
};

const create = async (
  ctx: restate.RpcContext,
  orderId: string,
  order: Order
) => {
  // Ignore already completed orders
  if (await completed(ctx)) {
    return
  }

  const { id, totalCost, deliveryDelay } = order;

  // 1. Set status
  ctx.send(orderstatus.service).setStatus(id, Status.CREATED);

  // 2. Handle payment
  const token = ctx.rand.uuidv4();
  const paid = await ctx.sideEffect(() => paymentClnt.charge(id, token, totalCost));

  if (!paid) {
    ctx.send(orderstatus.service).setStatus(id, Status.REJECTED);
    return;
  }

  // 3. Schedule preparation
  ctx.send(orderstatus.service).setStatus(id, Status.SCHEDULED);
  await ctx.sleep(deliveryDelay);

  // 4. Trigger preparation
  const preparationPromise = ctx.awakeable();
  await ctx.sideEffect(() => restaurant.prepare(id, preparationPromise.id));
  ctx.send(orderstatus.service).setStatus(id, Status.IN_PREPARATION);

  await preparationPromise.promise;
  ctx.send(orderstatus.service).setStatus(id, Status.SCHEDULING_DELIVERY);

  // 5. Find a driver and start delivery
  const deliveryId = ctx.rand.uuidv4();
  const deliveryPromise = ctx.awakeable<void>();
  await ctx.rpc(deliveryManager.service).start(deliveryId, {order, promise: deliveryPromise.id})
  await deliveryPromise.promise;
  ctx.send(orderstatus.service).setStatus(id, Status.DELIVERED);
};

export const router = restate.keyedRouter({
  create,
  eventHandler: restate.keyedEventHandler(eventHandler),
});

export type orderApi = typeof router;
export const service: restate.ServiceApi<orderApi> = { path: "order-workflow" };

File: python/end-to-end-applications/food-ordering/app/ordering/order_workflow.py
import uuid
from datetime import timedelta
from typing import TypedDict

from restate import Workflow, WorkflowContext, WorkflowSharedContext

from ordering.clients.payment_client import PaymentClient
from ordering.clients.restaurant_client import RestaurantClient
from ordering.types.types import Status
import ordering.delivery_manager as delivery_manager

payment_client = PaymentClient()
restaurant_client = RestaurantClient()

order_workflow = Workflow("order-workflow")


class Product(TypedDict):
    product_id: str
    description: str
    quantity: int


class Order(TypedDict):
    id: str
    restaurant_id: str
    products: list[Product]
    total_cost: int
    delivery_delay: int


@order_workflow.main()
async def run(ctx: WorkflowContext, order: Order):
    id, total_cost, delivery_delay = (
        ctx.key(),
        order["total_cost"],
        order["delivery_delay"],
    )

    ctx.set("status", Status.CREATED)

    token = str(ctx.uuid())
    paid = await ctx.run_typed("payment", payment_client.charge, token=token, amount=total_cost)

    if not paid:
        ctx.set("status", Status.REJECTED)
        return

    ctx.set("status", Status.SCHEDULED)
    await ctx.sleep(timedelta(milliseconds=delivery_delay))

    await ctx.run_typed("prepare", restaurant_client.prepare, order_id=id)
    ctx.set("status", Status.IN_PREPARATION)

    await ctx.promise("preparation_finished").value()
    ctx.set("status", Status.SCHEDULING_DELIVERY)

    delivery_id = str(ctx.uuid())
    ctx.object_send(delivery_manager.start, delivery_id, arg=order)

    await ctx.promise("driver_selected").value()
    ctx.set("status", Status.WAITING_FOR_DRIVER)
    await ctx.promise("driver_at_restaurant").value()
    ctx.set("status", Status.IN_DELIVERY)
    await ctx.promise("delivery_finished").value()
    ctx.set("status", Status.DELIVERED)


@order_workflow.handler(name="finishedPreparation")
async def finished_preparation(ctx: WorkflowSharedContext):
    await ctx.promise("preparation_finished").resolve(None)


@order_workflow.handler()
async def selected_driver(ctx: WorkflowSharedContext):
    await ctx.promise("driver_selected").resolve(None)


@order_workflow.handler()
async def signal_driver_at_restaurant(ctx: WorkflowSharedContext):
    await ctx.promise("driver_at_restaurant").resolve(None)


@order_workflow.handler()
async def signal_delivery_finished(ctx: WorkflowSharedContext):
    await ctx.promise("delivery_finished").resolve(None)


@order_workflow.handler()
async def get_status(ctx: WorkflowSharedContext):
    return await ctx.get("status")
